src/Categories/Category/Construction/KaroubiEnvelope.agda

```diff
@@ -0,0 +1,71 @@
+{-# OPTIONS --without-K --safe #-}
+
+open import Categories.Category.Core
+
+-- Karoubi Envelopes. These are the free completions of categories under split idempotents
+module Categories.Category.Construction.KaroubiEnvelope {o ℓ e} (𝒞 : Category o ℓ e) where
+
+open import Level
+
+open import Categories.Morphism.Idempotent 𝒞
+open import Categories.Morphism.Reasoning 𝒞
+
+open Category 𝒞
+open HomReasoning
+open Equiv
+
+-- It's nicer to work with the bundled form of idempotents here.
+-- We could use Σ types, but that gets a bit messy.
+record BundledIdempotent : Set (o ⊔ ℓ ⊔ e) where
+  field
+    {obj} : Obj
+    isIdempotent : Idempotent obj
+
+  open Idempotent isIdempotent public
+
+record Idempotent⇒ (I J : BundledIdempotent) : Set (ℓ ⊔ e) where
+  private
+    module I = BundledIdempotent I
+    module J = BundledIdempotent J
+  field
+    hom : I.obj ⇒ J.obj
+    absorbˡ : J.idem ∘ hom ≈ hom
+    absorbʳ : hom ∘ I.idem ≈ hom
+
+open BundledIdempotent
+open Idempotent⇒
+
+KaroubiEnvelope : Category (o ⊔ ℓ ⊔ e) (ℓ ⊔ e) e
+KaroubiEnvelope = record
+  { Obj = BundledIdempotent
+  ; _⇒_ = Idempotent⇒
+  ; _≈_ = λ f g → Idempotent⇒.hom f ≈ Idempotent⇒.hom g
+  ; id = λ {I} → record
```

Comment:
You may as well define `id` and composition in that same file too. That seems to be the consensus.